File: customfit-flutter-client-sdk/lib/src/core/util/config_cache.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';
import '../../logging/logger.dart';

/// Manages caching of configuration responses
/// This enables immediate access to cached configurations on startup
/// while still fetching updated configurations from the server.
class ConfigCache {
  static const String _configCacheKey = 'cf_cached_config_data';
  static const String _metadataCacheKey = 'cf_cached_config_metadata';

  /// Cache configuration data
  ///
  /// @param configMap The configuration map to cache
  /// @param lastModified The Last-Modified header value
  /// @param etag The ETag header value
  /// @return Future<bool> true if successfully cached, false otherwise
  Future<bool> cacheConfig(
    Map<String, dynamic> configMap,
    String? lastModified,
    String? etag,
  ) async {
    try {
      // Get shared prefs instance
      final prefs = await SharedPreferences.getInstance();

      // Serialize the config map to JSON
      final configJson = jsonEncode(configMap);

      // Store metadata separately
      final metadata = {
        'lastModified': lastModified ?? '',
        'etag': etag ?? '',
        'timestamp': DateTime.now().millisecondsSinceEpoch,
      };

      final metadataJson = jsonEncode(metadata);

      // Save both to shared preferences
      await prefs.setString(_configCacheKey, configJson);
      await prefs.setString(_metadataCacheKey, metadataJson);

      Logger.d('Configuration cached with ${configMap.length} entries');
      Logger.d(
          'Cached config metadata - Last-Modified: $lastModified, ETag: $etag');

      return true;
    } catch (e) {
      Logger.e('Error caching configuration: $e');
      return false;
    }
  }

  /// Get cached configuration data
  ///
  /// @return Future<Triple<Map<String, dynamic>?, String?, String?>> containing
  /// configuration map, Last-Modified value, and ETag value
  Future<ConfigCacheResult> getCachedConfig() async {
    try {
      // Get shared prefs instance
      final prefs = await SharedPreferences.getInstance();

      // Get cached config data
      final configJson = prefs.getString(_configCacheKey);
      final metadataJson = prefs.getString(_metadataCacheKey);

      if (configJson == null || metadataJson == null) {
        Logger.d('No cached configuration found');
        return ConfigCacheResult(null, null, null);
      }

      // Parse the data
      final configMap = jsonDecode(configJson) as Map<String, dynamic>;
      final metadata = jsonDecode(metadataJson) as Map<String, dynamic>;

      final lastModified = metadata['lastModified'] as String?;
      final etag = metadata['etag'] as String?;
      final timestamp = metadata['timestamp'] as int?;

      // Check if cache is still valid (24 hours)
      final now = DateTime.now().millisecondsSinceEpoch;
      final cacheAge = now - (timestamp ?? 0);
      const cacheTTL = 24 * 60 * 60 * 1000; // 24 hours in milliseconds

      if (cacheAge > cacheTTL) {
        Logger.d(
            'Cached configuration has expired (age: ${cacheAge / 1000 / 60 / 60} hours)');
        return ConfigCacheResult(null, null, null);
      }

      Logger.d('Found cached configuration with ${configMap.length} entries');
      Logger.d(
          'Cached config metadata - Last-Modified: $lastModified, ETag: $etag');
      Logger.d('Cache age: ${cacheAge / 1000 / 60} minutes');

      return ConfigCacheResult(configMap, lastModified, etag);
    } catch (e) {
      Logger.e('Error retrieving cached configuration: $e');
      return ConfigCacheResult(null, null, null);
    }
  }

  /// Clear cached configuration data
  Future<bool> clearCache() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.remove(_configCacheKey);
      await prefs.remove(_metadataCacheKey);

      Logger.d('Configuration cache cleared');
      return true;
    } catch (e) {
      Logger.e('Error clearing configuration cache: $e');
      return false;
    }
  }
}

/// Class to hold cache result values
class ConfigCacheResult {
  final Map<String, dynamic>? configMap;
  final String? lastModified;
  final String? etag;

  ConfigCacheResult(this.configMap, this.lastModified, this.etag);
}

File: customfit-kotlin-client-sdk/src/main/kotlin/customfit/ai/kotlinclient/client/managers/ConfigCache.kt
package customfit.ai.kotlinclient.client.managers

import customfit.ai.kotlinclient.logging.Timber
import customfit.ai.kotlinclient.utils.CacheManager
import customfit.ai.kotlinclient.utils.CachePolicy
import java.util.concurrent.atomic.AtomicReference

/**
 * Manages caching of configuration responses 
 * This enables immediate access to cached configurations on startup
 * while still fetching updated configurations from the server.
 */
class ConfigCache {
    companion object {
        private const val CONFIG_CACHE_KEY = "cf_config_data"
        private const val METADATA_CACHE_KEY = "cf_config_metadata"
        
        // Cache policy with 24 hour TTL but persisted across app restarts
        private val CONFIG_CACHE_POLICY = CachePolicy(
            ttlSeconds = 24 * 60 * 60, // 24 hours
            useStaleWhileRevalidate = true,
            evictOnAppRestart = false,
            persist = true
        )
    }
    
    private val cacheManager = CacheManager.getInstance()
    private val lastModifiedRef = AtomicReference<String?>(null)
    private val eTagRef = AtomicReference<String?>(null)
    
    /**
     * Store configuration data in cache
     * 
     * @param configMap The configuration map to cache
     * @param lastModified The Last-Modified header value
     * @param etag The ETag header value
     * @return True if successfully cached, false otherwise
     */
    fun cacheConfig(
        configMap: Map<String, Any>,
        lastModified: String?,
        etag: String?
    ): Boolean {
        // Store the config data
        val configResult = cacheManager.put(
            CONFIG_CACHE_KEY,
            configMap,
            CONFIG_CACHE_POLICY
        )
        
        // Store metadata separately
        val metadata = mapOf(
            "lastModified" to (lastModified ?: ""),
            "etag" to (etag ?: "")
        )
        
        val metadataResult = cacheManager.put(
            METADATA_CACHE_KEY,
            metadata,
            CONFIG_CACHE_POLICY
        )
        
        // Update in-memory refs
        lastModifiedRef.set(lastModified)
        eTagRef.set(etag)
        
        Timber.d("Configuration cached with Last-Modified: $lastModified, ETag: $etag")
        
        return configResult && metadataResult
    }
    
    /**
     * Get cached configuration data
     * 
     * @return Triple containing configuration map, Last-Modified value, and ETag value
     */
    fun getCachedConfig(): Triple<Map<String, Any>?, String?, String?> {
        // Retrieve config data with safe casting
        val configMap = try {
            val cachedData = cacheManager.get<Any>(CONFIG_CACHE_KEY)
            when (cachedData) {
                is Map<*, *> -> {
                    // Verify all keys are strings and values are of expected types
                    if (cachedData.keys.all { it is String }) {
                        @Suppress("UNCHECKED_CAST")
                        cachedData as Map<String, Any>
                    } else {
                        Timber.w("Cached config data has non-string keys, ignoring cache")
                        null
                    }
                }
                null -> null
                else -> {
                    Timber.w("Cached config data is not a Map (found ${cachedData::class.simpleName}), clearing cache")
                    cacheManager.remove(CONFIG_CACHE_KEY)
                    null
                }
            }
        } catch (e: Exception) {
            Timber.e(e, "Failed to retrieve cached config data: ${e.message}")
            // Clear corrupted cache
            cacheManager.remove(CONFIG_CACHE_KEY)
            null
        }
        
        // Get metadata if available with safe casting
        val metadata = try {
            val cachedMetadata = cacheManager.get<Any>(METADATA_CACHE_KEY)
            when (cachedMetadata) {
                is Map<*, *> -> {
                    // Verify all keys and values are strings
                    if (cachedMetadata.keys.all { it is String } && 
                        cachedMetadata.values.all { it is String }) {
                        @Suppress("UNCHECKED_CAST")
                        cachedMetadata as Map<String, String>
                    } else {
                        Timber.w("Cached metadata has invalid key/value types, ignoring")
                        null
                    }
                }
                null -> null
                else -> {
                    Timber.w("Cached metadata is not a Map (found ${cachedMetadata::class.simpleName}), clearing cache")
                    cacheManager.remove(METADATA_CACHE_KEY)
                    null
                }
            }
        } catch (e: Exception) {
            Timber.e(e, "Failed to retrieve cached metadata: ${e.message}")
            // Clear corrupted cache
            cacheManager.remove(METADATA_CACHE_KEY)
            null
        }
        
        val lastModified = metadata?.get("lastModified") ?: lastModifiedRef.get()
        val etag = metadata?.get("etag") ?: eTagRef.get()
        
        if (configMap != null) {
            Timber.d("Found cached configuration with ${configMap.size} entries")
            Timber.d("Cached config metadata - Last-Modified: $lastModified, ETag: $etag")
        } else {
            Timber.d("No cached configuration found")
        }
        
        return Triple(configMap, lastModified, etag)
    }
    
    /**
     * Clear cached configuration data
     */
    fun clearCache() {
        cacheManager.remove(CONFIG_CACHE_KEY)
        cacheManager.remove(METADATA_CACHE_KEY)
        lastModifiedRef.set(null)
        eTagRef.set(null)
        Timber.d("Configuration cache cleared")
    }
} 